internal/juju/credentials.go

```diff
@@ -0,0 +1,310 @@
+	cloudName := input.CloudName
+	controllerCredential := input.ControllerCredential
+	credentialName := input.Name
+
+	conn, err := c.GetConnection(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	client := cloudapi.NewClient(conn)
+	defer client.Close()
+
+	var clientCredentialFound jujucloud.Credential
+	if clientCredential {
+		existingCredentials, err := getExistingClientCredential(cloudName)
+		if err != nil {
+			return nil, err
+		}
+		clientCredentialFound = existingCredentials.AuthCredentials[credentialName]
+	}
+
+	var controllerCredentialFound jujucloud.Credential
+	if controllerCredential {
+		credentialContents, err := client.CredentialContents(cloudName, credentialName, true)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, content := range credentialContents {
+			if content.Error != nil {
+				continue
+			}
+			remoteCredential := content.Result.Content
+			if remoteCredential.Name == credentialName {
+				controllerCredentialFound = jujucloud.NewNamedCredential(
+					credentialName,
+					jujucloud.AuthType(remoteCredential.AuthType),
+					remoteCredential.Attributes,
+					false, //  CredentialContents does not provides this field
+				)
+				break
+			}
+		}
+	}
+
+	if controllerCredential && clientCredential {
+		// compare if they are the same
+		// lets just check auth_type for now
+		if clientCredentialFound.AuthType() != controllerCredentialFound.AuthType() {
+			return nil, fmt.Errorf("client and controller credentials have different auth type: %s, %s", clientCredentialFound.AuthType(), controllerCredentialFound.AuthType())
+		}
+	}
+
+	if controllerCredential {
+		return &ReadCredentialResponse{
+			CloudCredential: controllerCredentialFound,
+		}, nil
+	}
+
+	if clientCredential {
+		return &ReadCredentialResponse{
+			CloudCredential: clientCredentialFound,
+		}, nil
+	}
+
+	return nil, fmt.Errorf("credential %s not found for cloud %s", credentialName, cloudName)
+}
+
+func (c *credentialsClient) UpdateCredential(input UpdateCredentialInput) error {
+	cloudName := input.CloudName
+	credentialName := input.Name
+
+	conn, err := c.GetConnection(nil)
+	if err != nil {
+		return err
+	}
+
+	client := cloudapi.NewClient(conn)
+	defer client.Close()
+
+	currentUser := strings.TrimPrefix(conn.AuthTag().String(), PrefixUser)
+
+	cloudCredTag, err := GetCloudCredentialTag(cloudName, currentUser, credentialName)
+	if err != nil {
+		return err
+	}
+
+	cloudCredential := jujucloud.NewNamedCredential(
+		input.Name,
+		jujucloud.AuthType(input.AuthType),
+		input.Attributes,
+		false,
+	)
+
+	if input.ClientCredential {
+		if err := updateClientCredential(cloudName, credentialName, cloudCredential); err != nil {
+			return err
+		}
+	}
+
+	if input.ControllerCredential {
+		if _, err := client.UpdateCredentialsCheckModels(*cloudCredTag, cloudCredential); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func getExistingClientCredential(cloudName string) (*jujucloud.CloudCredential, error) {
```

Comment:
Great. This will be used only if the "client_credential" is set to true, by default is false and the credential will be added only to the controller.

Reply:
BTW in case of not finding the Juju client store (~/.local/share/juju/) this error will be returned:  "Error: credential not found for cloud microk8s: credentials for cloud microk8s not found"
The "credentials for cloud microk8s not found" is generated by "store.CredentialForCloud(cloudName)"

Reply:
Since there is no action, I'll set this as resolved but feel free to reopen it if needed.